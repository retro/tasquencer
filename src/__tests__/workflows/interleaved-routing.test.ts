import { Effect } from 'effect';
import { it } from 'vitest';

import { Builder, IdGenerator, Service } from '../../index.js';
import { getEnabledTaskNames, makeIdGenerator } from '../shared.js';

it('supports the interleaved routing pattern', ({ expect }) => {
  const workflowDefinition = Builder.workflow()
    .withName('interleaved-parallel')
    .startCondition('start')
    .condition('mutex')
    .task('initial_task', Builder.emptyTask())
    .task(
      'task_a',
      Builder.emptyTask().withSplitType('and').withJoinType('and')
    )
    .task(
      'task_b',
      Builder.emptyTask().withSplitType('and').withJoinType('and')
    )
    .task(
      'task_c',
      Builder.emptyTask().withSplitType('and').withJoinType('and')
    )
    .task(
      'task_d',
      Builder.emptyTask().withSplitType('and').withJoinType('and')
    )
    .task('finish_task', Builder.emptyTask().withJoinType('and'))
    .endCondition('end')
    .connectCondition('start', (to) => to.task('initial_task'))
    .connectTask('initial_task', (to) =>
      to.condition('mutex').task('task_a').task('task_c')
    )
    .connectTask('task_a', (to) => to.task('task_b').condition('mutex'))
    .connectTask('task_b', (to) => to.task('finish_task').condition('mutex'))
    .connectTask('task_c', (to) => to.task('task_d').condition('mutex'))
    .connectTask('task_d', (to) => to.task('finish_task').condition('mutex'))
    .connectTask('finish_task', (to) => to.condition('end'))
    .connectCondition('mutex', (to) =>
      to.task('task_a').task('task_b').task('task_c').task('task_d')
    )
    .cancellationRegion('finish_task', { conditions: ['mutex'] });

  const program = Effect.gen(function* ($) {
    const idGenerator = makeIdGenerator();

    const service = yield* $(
      workflowDefinition.build(),
      Effect.flatMap((workflow) => Service.initialize(workflow)),
      Effect.provideService(IdGenerator, idGenerator)
    );

    expect(yield* $(service.getState())).toMatchSnapshot();

    yield* $(service.start());
    const state1 = yield* $(service.getState());
    expect(state1).toMatchSnapshot();
    expect(getEnabledTaskNames(state1)).toEqual(new Set(['initial_task']));

    yield* $(service.startTask('initial_task'));
    const state2 = yield* $(service.getState());
    expect(state2).toMatchSnapshot();
    expect(getEnabledTaskNames(state2)).toEqual(new Set(['task_a', 'task_c']));

    yield* $(service.startTask('task_a'));
    const state3 = yield* $(service.getState());
    expect(state3).toMatchSnapshot();
    expect(getEnabledTaskNames(state3)).toEqual(new Set(['task_c', 'task_b']));

    yield* $(service.startTask('task_b'));
    const state4 = yield* $(service.getState());
    expect(state4).toMatchSnapshot();
    expect(getEnabledTaskNames(state4)).toEqual(new Set(['task_c']));

    yield* $(service.startTask('task_c'));
    const state5 = yield* $(service.getState());
    expect(state5).toMatchSnapshot();
    expect(getEnabledTaskNames(state5)).toEqual(new Set(['task_d']));

    yield* $(service.startTask('task_d'));
    const state6 = yield* $(service.getState());
    expect(state6).toMatchSnapshot();
    expect(getEnabledTaskNames(state6)).toEqual(new Set(['finish_task']));

    yield* $(service.startTask('finish_task'));
    const state7 = yield* $(service.getState());
    expect(state7).toMatchSnapshot();
    expect(state7.workflows[0]?.state).toBe('completed');
  });

  Effect.runSync(program);
});
