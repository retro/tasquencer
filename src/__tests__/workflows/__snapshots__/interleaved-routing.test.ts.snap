// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`supports the interleaved routing pattern 1`] = `
{
  "conditions": [
    {
      "marking": 0,
      "name": "start",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
    {
      "marking": 0,
      "name": "mutex",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
    {
      "marking": 0,
      "name": "end",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
    {
      "marking": 0,
      "name": "implicit:initial_task->task_a",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
    {
      "marking": 0,
      "name": "implicit:initial_task->task_c",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
    {
      "marking": 0,
      "name": "implicit:task_a->task_b",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
    {
      "marking": 0,
      "name": "implicit:task_b->finish_task",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
    {
      "marking": 0,
      "name": "implicit:task_c->task_d",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
    {
      "marking": 0,
      "name": "implicit:task_d->finish_task",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
  ],
  "tasks": [
    {
      "generation": 0,
      "name": "initial_task",
      "state": "disabled",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
    {
      "generation": 0,
      "name": "task_a",
      "state": "disabled",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
    {
      "generation": 0,
      "name": "task_b",
      "state": "disabled",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
    {
      "generation": 0,
      "name": "task_c",
      "state": "disabled",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
    {
      "generation": 0,
      "name": "task_d",
      "state": "disabled",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
    {
      "generation": 0,
      "name": "finish_task",
      "state": "disabled",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
  ],
  "workItems": [],
  "workflows": [
    {
      "context": undefined,
      "id": "workflow-1",
      "name": "interleaved-parallel",
      "parent": null,
      "state": "initialized",
    },
  ],
}
`;

exports[`supports the interleaved routing pattern 2`] = `
{
  "conditions": [
    {
      "marking": 1,
      "name": "start",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
    {
      "marking": 0,
      "name": "mutex",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
    {
      "marking": 0,
      "name": "end",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
    {
      "marking": 0,
      "name": "implicit:initial_task->task_a",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
    {
      "marking": 0,
      "name": "implicit:initial_task->task_c",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
    {
      "marking": 0,
      "name": "implicit:task_a->task_b",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
    {
      "marking": 0,
      "name": "implicit:task_b->finish_task",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
    {
      "marking": 0,
      "name": "implicit:task_c->task_d",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
    {
      "marking": 0,
      "name": "implicit:task_d->finish_task",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
  ],
  "tasks": [
    {
      "generation": 0,
      "name": "initial_task",
      "state": "enabled",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
    {
      "generation": 0,
      "name": "task_a",
      "state": "disabled",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
    {
      "generation": 0,
      "name": "task_b",
      "state": "disabled",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
    {
      "generation": 0,
      "name": "task_c",
      "state": "disabled",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
    {
      "generation": 0,
      "name": "task_d",
      "state": "disabled",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
    {
      "generation": 0,
      "name": "finish_task",
      "state": "disabled",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
  ],
  "workItems": [],
  "workflows": [
    {
      "context": undefined,
      "id": "workflow-1",
      "name": "interleaved-parallel",
      "parent": null,
      "state": "started",
    },
  ],
}
`;

exports[`supports the interleaved routing pattern 3`] = `
{
  "conditions": [
    {
      "marking": 0,
      "name": "start",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
    {
      "marking": 1,
      "name": "mutex",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
    {
      "marking": 0,
      "name": "end",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
    {
      "marking": 1,
      "name": "implicit:initial_task->task_a",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
    {
      "marking": 1,
      "name": "implicit:initial_task->task_c",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
    {
      "marking": 0,
      "name": "implicit:task_a->task_b",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
    {
      "marking": 0,
      "name": "implicit:task_b->finish_task",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
    {
      "marking": 0,
      "name": "implicit:task_c->task_d",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
    {
      "marking": 0,
      "name": "implicit:task_d->finish_task",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
  ],
  "tasks": [
    {
      "generation": 1,
      "name": "initial_task",
      "state": "completed",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
    {
      "generation": 0,
      "name": "task_a",
      "state": "enabled",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
    {
      "generation": 0,
      "name": "task_b",
      "state": "disabled",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
    {
      "generation": 0,
      "name": "task_c",
      "state": "enabled",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
    {
      "generation": 0,
      "name": "task_d",
      "state": "disabled",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
    {
      "generation": 0,
      "name": "finish_task",
      "state": "disabled",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
  ],
  "workItems": [],
  "workflows": [
    {
      "context": undefined,
      "id": "workflow-1",
      "name": "interleaved-parallel",
      "parent": null,
      "state": "started",
    },
  ],
}
`;

exports[`supports the interleaved routing pattern 4`] = `
{
  "conditions": [
    {
      "marking": 0,
      "name": "start",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
    {
      "marking": 1,
      "name": "mutex",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
    {
      "marking": 0,
      "name": "end",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
    {
      "marking": 0,
      "name": "implicit:initial_task->task_a",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
    {
      "marking": 1,
      "name": "implicit:initial_task->task_c",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
    {
      "marking": 1,
      "name": "implicit:task_a->task_b",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
    {
      "marking": 0,
      "name": "implicit:task_b->finish_task",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
    {
      "marking": 0,
      "name": "implicit:task_c->task_d",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
    {
      "marking": 0,
      "name": "implicit:task_d->finish_task",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
  ],
  "tasks": [
    {
      "generation": 1,
      "name": "initial_task",
      "state": "completed",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
    {
      "generation": 1,
      "name": "task_a",
      "state": "completed",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
    {
      "generation": 0,
      "name": "task_b",
      "state": "enabled",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
    {
      "generation": 0,
      "name": "task_c",
      "state": "enabled",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
    {
      "generation": 0,
      "name": "task_d",
      "state": "disabled",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
    {
      "generation": 0,
      "name": "finish_task",
      "state": "disabled",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
  ],
  "workItems": [],
  "workflows": [
    {
      "context": undefined,
      "id": "workflow-1",
      "name": "interleaved-parallel",
      "parent": null,
      "state": "started",
    },
  ],
}
`;

exports[`supports the interleaved routing pattern 5`] = `
{
  "conditions": [
    {
      "marking": 0,
      "name": "start",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
    {
      "marking": 1,
      "name": "mutex",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
    {
      "marking": 0,
      "name": "end",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
    {
      "marking": 0,
      "name": "implicit:initial_task->task_a",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
    {
      "marking": 1,
      "name": "implicit:initial_task->task_c",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
    {
      "marking": 0,
      "name": "implicit:task_a->task_b",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
    {
      "marking": 1,
      "name": "implicit:task_b->finish_task",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
    {
      "marking": 0,
      "name": "implicit:task_c->task_d",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
    {
      "marking": 0,
      "name": "implicit:task_d->finish_task",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
  ],
  "tasks": [
    {
      "generation": 1,
      "name": "initial_task",
      "state": "completed",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
    {
      "generation": 1,
      "name": "task_a",
      "state": "completed",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
    {
      "generation": 1,
      "name": "task_b",
      "state": "completed",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
    {
      "generation": 0,
      "name": "task_c",
      "state": "enabled",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
    {
      "generation": 0,
      "name": "task_d",
      "state": "disabled",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
    {
      "generation": 0,
      "name": "finish_task",
      "state": "disabled",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
  ],
  "workItems": [],
  "workflows": [
    {
      "context": undefined,
      "id": "workflow-1",
      "name": "interleaved-parallel",
      "parent": null,
      "state": "started",
    },
  ],
}
`;

exports[`supports the interleaved routing pattern 6`] = `
{
  "conditions": [
    {
      "marking": 0,
      "name": "start",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
    {
      "marking": 1,
      "name": "mutex",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
    {
      "marking": 0,
      "name": "end",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
    {
      "marking": 0,
      "name": "implicit:initial_task->task_a",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
    {
      "marking": 0,
      "name": "implicit:initial_task->task_c",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
    {
      "marking": 0,
      "name": "implicit:task_a->task_b",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
    {
      "marking": 1,
      "name": "implicit:task_b->finish_task",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
    {
      "marking": 1,
      "name": "implicit:task_c->task_d",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
    {
      "marking": 0,
      "name": "implicit:task_d->finish_task",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
  ],
  "tasks": [
    {
      "generation": 1,
      "name": "initial_task",
      "state": "completed",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
    {
      "generation": 1,
      "name": "task_a",
      "state": "completed",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
    {
      "generation": 1,
      "name": "task_b",
      "state": "completed",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
    {
      "generation": 1,
      "name": "task_c",
      "state": "completed",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
    {
      "generation": 0,
      "name": "task_d",
      "state": "enabled",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
    {
      "generation": 0,
      "name": "finish_task",
      "state": "disabled",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
  ],
  "workItems": [],
  "workflows": [
    {
      "context": undefined,
      "id": "workflow-1",
      "name": "interleaved-parallel",
      "parent": null,
      "state": "started",
    },
  ],
}
`;

exports[`supports the interleaved routing pattern 7`] = `
{
  "conditions": [
    {
      "marking": 0,
      "name": "start",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
    {
      "marking": 1,
      "name": "mutex",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
    {
      "marking": 0,
      "name": "end",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
    {
      "marking": 0,
      "name": "implicit:initial_task->task_a",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
    {
      "marking": 0,
      "name": "implicit:initial_task->task_c",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
    {
      "marking": 0,
      "name": "implicit:task_a->task_b",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
    {
      "marking": 1,
      "name": "implicit:task_b->finish_task",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
    {
      "marking": 0,
      "name": "implicit:task_c->task_d",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
    {
      "marking": 1,
      "name": "implicit:task_d->finish_task",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
  ],
  "tasks": [
    {
      "generation": 1,
      "name": "initial_task",
      "state": "completed",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
    {
      "generation": 1,
      "name": "task_a",
      "state": "completed",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
    {
      "generation": 1,
      "name": "task_b",
      "state": "completed",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
    {
      "generation": 1,
      "name": "task_c",
      "state": "completed",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
    {
      "generation": 1,
      "name": "task_d",
      "state": "completed",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
    {
      "generation": 0,
      "name": "finish_task",
      "state": "enabled",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
  ],
  "workItems": [],
  "workflows": [
    {
      "context": undefined,
      "id": "workflow-1",
      "name": "interleaved-parallel",
      "parent": null,
      "state": "started",
    },
  ],
}
`;

exports[`supports the interleaved routing pattern 8`] = `
{
  "conditions": [
    {
      "marking": 0,
      "name": "start",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
    {
      "marking": 0,
      "name": "mutex",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
    {
      "marking": 1,
      "name": "end",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
    {
      "marking": 0,
      "name": "implicit:initial_task->task_a",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
    {
      "marking": 0,
      "name": "implicit:initial_task->task_c",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
    {
      "marking": 0,
      "name": "implicit:task_a->task_b",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
    {
      "marking": 0,
      "name": "implicit:task_b->finish_task",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
    {
      "marking": 0,
      "name": "implicit:task_c->task_d",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
    {
      "marking": 0,
      "name": "implicit:task_d->finish_task",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
  ],
  "tasks": [
    {
      "generation": 1,
      "name": "initial_task",
      "state": "completed",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
    {
      "generation": 1,
      "name": "task_a",
      "state": "completed",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
    {
      "generation": 1,
      "name": "task_b",
      "state": "completed",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
    {
      "generation": 1,
      "name": "task_c",
      "state": "completed",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
    {
      "generation": 1,
      "name": "task_d",
      "state": "completed",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
    {
      "generation": 1,
      "name": "finish_task",
      "state": "completed",
      "workflowId": "workflow-1",
      "workflowName": "interleaved-parallel",
    },
  ],
  "workItems": [],
  "workflows": [
    {
      "context": undefined,
      "id": "workflow-1",
      "name": "interleaved-parallel",
      "parent": null,
      "state": "completed",
    },
  ],
}
`;
